File: backend/app/rules.py
from __future__ import annotations

import re
from typing import List, Tuple

# Precompiled regex patterns for common XSS and SQLi
XSS_PATTERNS = [
    re.compile(r"<\s*script\b", re.IGNORECASE),
    re.compile(r"on\w+\s*=\s*\\?\"?[^\s>]+", re.IGNORECASE),  # onerror=, onclick=
    re.compile(r"javascript:\s*", re.IGNORECASE),
    re.compile(r"<\s*img\b[^>]*src\s*=", re.IGNORECASE),
]

SQLI_PATTERNS = [
    re.compile(r"\bUNION\s+SELECT\b", re.IGNORECASE),
    re.compile(r"\bDROP\s+TABLE\b", re.IGNORECASE),
    re.compile(r"\bOR\s+1=1\b", re.IGNORECASE),
    re.compile(r"\b--\s", re.IGNORECASE),
    re.compile(r";\s*SHUTDOWN\b", re.IGNORECASE),
]


# Add patterns for command-line tools and suspicious URLs
GENERIC_DANGEROUS = [
    re.compile(r"<\s*/?\s*iframe\b", re.IGNORECASE),
    re.compile(r"<\s*svg\b", re.IGNORECASE),
    re.compile(r"\bexec\b|\bxp_cmdshell\b", re.IGNORECASE),
    re.compile(r"\bcurl\b", re.IGNORECASE),
    re.compile(r"\bwget\b", re.IGNORECASE),
    re.compile(r"http[s]?://[\w\.-]+", re.IGNORECASE),
    re.compile(r"exploit=\d+", re.IGNORECASE),
    # Path traversal
    re.compile(r"\.\./+", re.IGNORECASE),
    re.compile(r"/etc/passwd", re.IGNORECASE),
    # Command injection
    re.compile(r"\|\s*rm\s+-rf\s+/", re.IGNORECASE),
    re.compile(r";\s*rm\s+-rf\s+/", re.IGNORECASE),
    # PHP RCE
    re.compile(r"<\?php.*system\s*\(", re.IGNORECASE),
    re.compile(r"\$_GET\s*\[\s*['\"]cmd['\"]\s*\]", re.IGNORECASE),
]


def scan_rules(payload: str, sensitivity: str) -> Tuple[List[str], str]:
    """Return (rule_hits, severity) based on rule matches and sensitivity.

    severity in {Clean, Suspicious, High Threat}
    """
    hits: List[str] = []

    for pat in XSS_PATTERNS:
        if pat.search(payload):
            hits.append("XSS:" + pat.pattern)

    for pat in SQLI_PATTERNS:
        if pat.search(payload):
            hits.append("SQLi:" + pat.pattern)

    for pat in GENERIC_DANGEROUS:
        if pat.search(payload):
            hits.append("GENERIC:" + pat.pattern)

    # Sensitivity tuning
    sensitivity = (sensitivity or "Medium").capitalize()
    num_hits = len(hits)

    if num_hits == 0:
        return hits, "Clean"

    if sensitivity == "Low":
        # Only heavy patterns escalate high threat
        if any(h.startswith("SQLi:") for h in hits) or any("<\\s*script" in h for h in hits):
            return hits, "High Threat"
        return hits, "Suspicious"

    if sensitivity == "Medium":
        # A couple hits => suspicious, SQLi => high
        if any(h.startswith("SQLi:") for h in hits):
            return hits, "High Threat"
        if num_hits >= 2:
            return hits, "Suspicious"
        return hits, "Suspicious"

    # Paranoid
    if any(h.startswith("SQLi:") for h in hits):
        return hits, "High Threat"
    if num_hits >= 1:
        return hits, "Suspicious"

    return hits, "Clean"

File: frontend/lib/rules.ts
// Precompiled regex patterns for common XSS and SQLi
const XSS_PATTERNS = [
  /< *script\b/i,
  /on\w+ *= *\\?"?[^\s>]+/i, // onerror=, onclick=
  /javascript: */i,
  /< *img\b[^>]*src *=/i,
];

const SQLI_PATTERNS = [
  /\bUNION +SELECT\b/i,
  /\bDROP +TABLE\b/i,
  /\bOR +1=1\b/i,
  /\b-- /i,
  /; *SHUTDOWN\b/i,
];

// Add patterns for command-line tools and suspicious URLs
const GENERIC_DANGEROUS = [
  /< *\/? *iframe\b/i,
  /< *svg\b/i,
  /\bexec\b|\bxp_cmdshell\b/i,
  /\bcurl\b/i,
  /\bwget\b/i,
  /https?:\/\/[\w\.-]+/i,
  /exploit=\d+/i,
  // Path traversal
  /\.\.\//i,
  /\/etc\/passwd/i,
  // Command injection
  /\| *rm +-rf +\//i,
  /; *rm +-rf +\//i,
  // PHP RCE
  /<\?php.*system *\(/i,
  /\$_GET *\[ *['"]cmd['"] *\]/i,
];

export function scanRules(
  payload: string,
  sensitivity: string
): [string[], string] {
  const hits: string[] = [];

  for (const pattern of XSS_PATTERNS) {
    if (pattern.test(payload)) {
      hits.push(`XSS:${pattern.source}`);
    }
  }

  for (const pattern of SQLI_PATTERNS) {
    if (pattern.test(payload)) {
      hits.push(`SQLi:${pattern.source}`);
    }
  }

  for (const pattern of GENERIC_DANGEROUS) {
    if (pattern.test(payload)) {
      hits.push(`GENERIC:${pattern.source}`);
    }
  }

  // Sensitivity tuning
  const normalizedSensitivity =
    (sensitivity || "Medium").charAt(0).toUpperCase() +
    sensitivity.slice(1).toLowerCase();
  const numHits = hits.length;

  if (numHits === 0) {
    return [hits, "Clean"];
  }

  if (normalizedSensitivity === "Low") {
    // Only heavy patterns escalate high threat
    if (
      hits.some((h) => h.startsWith("SQLi:")) ||
      hits.some((h) => h.includes("< *script"))
    ) {
      return [hits, "High Threat"];
    }
    return [hits, "Suspicious"];
  }

  if (normalizedSensitivity === "Medium") {
    // A couple hits => suspicious, SQLi => high
    if (hits.some((h) => h.startsWith("SQLi:"))) {
      return [hits, "High Threat"];
    }
    if (numHits >= 2) {
      return [hits, "Suspicious"];
    }
    return [hits, "Suspicious"];
  }

  // Paranoid
  if (hits.some((h) => h.startsWith("SQLi:"))) {
    return [hits, "High Threat"];
  }
  if (numHits >= 1) {
    return [hits, "Suspicious"];
  }

  return [hits, "Clean"];
}
